BJS2-66374-request-for-analytics

gulnaz21 opened the pull request.

gulnaz21 committed:
BJS2-66374-request-for-analytics

Zhltsk-V commented on src/main/java/faang/school/analytics/controller/AnalyticsController.java:
```diff
@@ -0,0 +1,85 @@
+@RequiredArgsConstructor
+@Slf4j
+@Tag(
+        name = "API аналитики",
```
Мы договаривались писать swagger на английском языке, перепиши пожалуйста)

Zhltsk-V commented on src/main/java/faang/school/analytics/controller/AnalyticsController.java:
```diff
@@ -0,0 +1,85 @@
+        return analyticsService.getAnalytics(receiverId, eventType, fromDate, toDate);
+    }
+
+    public void validateDateParametersAndInterval(Integer intervalId, LocalDateTime fromDate, LocalDateTime toDate) {
```
Контроллер не должен заниматься валидацией (только если первичной), давай лучше сделаем это в сервисе

gulnaz21 replied:
По сути это и есть первичная валидация, я же просто проверяю, что значения не null

Zhltsk-V commented on src/main/java/faang/school/analytics/dto/AnalyticsEventDto.java:
```diff
@@ -0,0 +1,30 @@
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@Schema(description = "DTO для представления события аналитики")
```
Swagger на английском =)

Zhltsk-V commented on src/main/java/faang/school/analytics/model/Interval.java:
```diff
@@ -0,0 +1,49 @@
+                return interval;
+            }
+        }
+        throw new IllegalArgumentException("Unknown event type: " + type);
```
Можно добавить логгирование

gulnaz21 committed:
fix:
  - сваггер переписан на английский
  - добавлено логирование

RinatBeybutov commented on src/main/java/faang/school/analytics/dto/AnalyticsEventDto.java:
```diff
@@ -0,0 +1,30 @@
+    private long receiverId;
+
+    @Schema(description = "Event initiator ID", example = "2002")
+    private long actorId;
```
здесь обертки везде

RinatBeybutov commented on src/main/java/faang/school/analytics/model/AnalyticsEvent.java:
```diff
@@ -17,28 +17,47 @@
+ * Сущность, представляющая событие аналитики в системе.
+ * Хранит информацию о событиях, их участниках и времени возникновения.
+ */
 @Data
```
дата в сущности - это ружье, которое потом может выстрелить. А ты в это время будешь его чистить

RinatBeybutov commented on src/main/java/faang/school/analytics/model/AnalyticsEvent.java:
```diff
@@ -17,28 +17,47 @@
+     */
     @Id
     @GeneratedValue(strategy = GenerationType.IDENTITY)
     private long id;
```
Нужно поменять на обертки, примитивы в сущностях - плохая практика

gulnaz21 replied:
А нужна ли эта обертка, если эти значения не должны быть null?

RinatBeybutov commented on src/main/java/faang/school/analytics/service/AnalyticsService.java:
```diff
@@ -0,0 +1,46 @@
+     * @return список {@link AnalyticsEventDto}
+     */
+    @Transactional(readOnly = true)
+    public List<AnalyticsEventDto> getAnalytics(long receiverId, EventType eventType,
```
receiverId обертка должна быть?

gulnaz21 replied:
я использовала примитив long чтобы подчеркнуть, что receiverId не может быть null, и мы ожидаем уже валидное значение на уровне сервиса
В таком случае стоит ли мне менять на обертку или достаточно проверки на null в контроллере?

RinatBeybutov replied:
лучше не переходить на примитивы

RinatBeybutov commented on src/main/java/faang/school/analytics/service/AnalyticsService.java:
```diff
@@ -0,0 +1,46 @@
+        Stream<AnalyticsEvent> events = analyticsEventRepository.findByReceiverIdAndEventType(receiverId, eventType);
+        return events.filter(event ->
+                        event.getReceivedAt().isAfter(fromDate) &&
+                                event.getReceivedAt().isBefore(toDate))
```
давай сделаем метод boolean isReceivedAtBetween(LocalDateTime fromDate, LocalDateTime toDate) в классе AnalyticsEvent. Так лаконичнее будет, чем получать getReceivedAt()

gulnaz21 committed:
fix:
  - добавлена обработка ошибки MissingServletRequestParameterException
  - Удалена аннотация @Data в AnalyticsEvent, примитивный long заменен на обертку Long

gulnaz21 committed:
fix:
  - В AnalyticsService примитивный long заменен на обертку Long
